promql/prettier/prettier.go

```diff
@@ -0,0 +1,480 @@
+					} else if !(itemp >= keywordIndex && itemp <= closingParenIndex) {
+						tempItems = append(tempItems, items[itemp])
+					}
+				}
+				items = tempItems
+			}
+
+		case parser.IDENTIFIER:
+			//	TODO: currently its assumed that STRING are always label values.
+			itr := advanceComments(items, i)
+			if i < 1 || item.Val != "__name__" || items[itr+2].Typ != parser.STRING {
+				continue
+			}
+			var (
+				leftBraceIndex  = -1
+				rightBraceIndex = -1
+				labelValItem    = items[itr+2]
+				metricName      = labelValItem.Val[1 : len(labelValItem.Val)-1] // Trim inverted commas.
+				tmp             []parser.Item
+				skipBraces      bool // For handling metric_name{} -> metric_name
+			)
+			// We aim to convert __name__ to an ideal metric_name if the value of that labels is atomic.
+			// Since a non-atomic metric_name will contain alphabets other than a-z and A-Z including _,
+			// anything that violates this ceases the formatting of that particular label item.
+			// If this is not done then the output from the prettier might be an un-parsable expression.
+			if !regexp.MustCompile("^[a-z_A-Z]+$").MatchString(metricName) {
+				continue
+			}
+			for backScanIndex := itr; backScanIndex >= 0; backScanIndex-- {
+				if items[backScanIndex].Typ == parser.LEFT_BRACE {
+					leftBraceIndex = backScanIndex
+					break
+				}
+			}
+			for forwardScanIndex := itr; forwardScanIndex < len(items); forwardScanIndex++ {
+				if items[forwardScanIndex].Typ == parser.RIGHT_BRACE {
+					rightBraceIndex = forwardScanIndex
+					break
+				}
+			}
+			// TODO: assuming comments are not present at this place.
+			itr = advanceComments(items, itr)
+			if items[itr+3].Typ == parser.COMMA {
+				skipBraces = rightBraceIndex-5 == advanceComments(items, leftBraceIndex)
+			} else {
+				skipBraces = rightBraceIndex-4 == advanceComments(items, leftBraceIndex)
+			}
+			identifierItem := parser.Item{Typ: parser.IDENTIFIER, Val: metricName, Pos: 0}
+			for j := 0; j < len(items); j++ {
+				if j <= rightBraceIndex && j >= leftBraceIndex {
+					if j == leftBraceIndex {
+						tmp = append(tmp, identifierItem)
+						if !skipBraces {
+							tmp = append(tmp, items[j])
+						}
+						continue
+					} else if items[itr+3].Typ == parser.COMMA && j == itr+3 || j >= itr && j < itr+3 {
+						continue
+					}
+					if skipBraces && (items[j].Typ == parser.LEFT_BRACE || items[j].Typ == parser.RIGHT_BRACE) {
+						continue
+					}
+				}
+				tmp = append(tmp, items[j])
+			}
+			items = tmp
+		}
+	}
+	return p.refreshLexItems(items)
+}
+
+func (p *Prettier) prettify(items []parser.Item, index int, result string) (string, error) {
+	if items[index].Typ == parser.EOF {
+		return result, nil
+	}
+	var (
+		item       = items[index]
+		nodeInfo   = newNodeInfo(p.Node, 100, item)
+		node       = nodeInfo.getNode(p.Node, item)
+		currNode   = newNodeInfo(node, 100, item)
+		baseIndent int
+	)
+
+	if reflect.TypeOf(node).String() == "*parser.BinaryExpr" {
```

Comment:
It's better to use a type switch or checked cast here, since that adds more compile time safety.

Generally, the use of the `reflect` package should be reduced as much as possible, since it is quite error prone.

Reply:
I did proceed in that way, but then felt that using switch for a single type check looked more than required. Still, should I replace with switch here?

Reply:
Yes.